ffsync/rds_locker.go

```diff
@@ -0,0 +1,174 @@
+package ffsync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/featureform/fferr"
+	"github.com/featureform/helpers"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v4/pgxpool"
+	_ "github.com/lib/pq"
+)
+
+const (
+	lock_expiration_time = "5 minutes"
+	key_length           = 2048
+)
+
+type rdsKey struct {
+	id   string
+	key  string
+	done chan error
+}
+
+func (k rdsKey) ID() string {
+	return k.id
+}
+
+func (k rdsKey) Key() string {
+	return k.key
+}
+
+func NewRDSLocker(config helpers.RDSConfig) (Locker, error) {
+	const tableName = "ff_locks"
+
+	db, err := helpers.NewRDSPoolConnection(config)
+	if err != nil {
+		return nil, err
+	}
+
+	connection, err := db.Acquire(context.Background())
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to acquire connection from the database pool: %w", err))
+	}
+
+	err = connection.Ping(context.Background())
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to ping the database: %w", err))
+	}
+
+	// Create a table to store the locks
+	tableCreationSQL := createTableQuery(tableName)
+	_, err = db.Exec(context.Background(), tableCreationSQL)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to create table %s: %w", tableName, err))
+	}
+
+	return &rdsLocker{
+		db:         db,
+		tableName:  tableName,
+		connection: connection,
+	}, nil
+}
+
+type rdsLocker struct {
+	db         *pgxpool.Pool
+	tableName  string
+	connection *pgxpool.Conn
+}
+
+func (l *rdsLocker) Lock(key string) (Key, error) {
+	if key == "" {
+		return nil, fferr.NewInternalError(fmt.Errorf("cannot lock an empty key"))
+	} else if len(key) > key_length {
+		return nil, fferr.NewInternalError(fmt.Errorf("key is too long: %d, max length: %d", len(key), key_length))
+	}
+
+	id := uuid.New().String()
+
+	// Get the lock
+	lockQuery := l.lockQuery()
+	r, err := l.db.Exec(context.Background(), lockQuery, id, key)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to lock key %s: %v", key, err))
+	} else if r.RowsAffected() == 0 {
+		return nil, fferr.NewKeyAlreadyLockedError(key, id, nil)
+	}
+
+	done := make(chan error)
+	lockKey := rdsKey{
+		id:   id,
+		key:  key,
+		done: done,
+	}
+
+	go l.updateLockTime(&lockKey)
+
+	return lockKey, nil
+}
+
+func (l *rdsLocker) updateLockTime(key *rdsKey) {
+	ticker := time.NewTicker(UpdateSleepTime)
```

Comment:
Why is this not a const? And should we check that it's less than half the lock_experation_time?